app/scripts/services/authorization.js

```diff
@@ -0,0 +1,121 @@
+'use strict';
+
+angular.module("openshiftConsole")
+  .factory("AuthorizationService", function($q, $cacheFactory, Logger, $interval, DataService){
+    
+    var currentProject = null;
+    var cachedRulesByProject = $cacheFactory('rulesCache', {
+          number: 10
+        });
+
+    var REVIEW_RESOURCES = ["localresourceaccessreviews", "localsubjectaccessreviews", "resourceaccessreviews", "selfsubjectrulesreviews", "subjectaccessreviews"];
+
+    // Transform data from:
+    // rules = {resources: ["jobs"], apiGroups: ["extensions"], verbs:["create","delete","get","list","update"]}
+    // into:
+    // normalizedRules = {"extensions/jobs": ["create","delete","get","list","update"]}
+    var normalizeRules = function(rules) {
+      var normalizedRules = {};
+
+      _.each(rules, function(rule) {
+        _.each(rule.resources, function(resource) {
+          _.each(rule.apiGroups, function(apiGroup) {
+            normalizedRules[(apiGroup === "") ? resource : apiGroup + "/" + resource] = rule.verbs;
+          });
+        });
+      });
+      return normalizedRules;
+    };
+
+    // Check if resource name meets one of following conditions, since those resources can't be create/update via `Add to project` page:
+    //  - 'projectrequests'
+    //  - subresource that contains '/', eg: 'builds/source', 'builds/logs', ...
+    //  - resource is in REVIEW_RESOURCES list 
+    var checkResource = function(resource) {
+      if (resource === "projectrequests" || _.contains(resource, "/") || _.contains(REVIEW_RESOURCES, resource)) {
+        return false;
+      } else {
+        return true;
+      }
+    };
+
+    // Check if user can create/update any resource on the 'Add to project' so the button will be displayed.
+    var canAddToProject = function(rules) {
+      var canAddToProject = null;
+      _.each(rules, function(rule) {
+        _.each(rule.resources, function(resource) {
+          if (!canAddToProject && checkResource(resource) && (!_.isEmpty(_.intersection(rule.verbs ,(["*", "create", "update"]))))) {
+            canAddToProject = true;
```

Comment:
i think this would be better if it were nested _.some so that it shortcircuits as fast as possible

```
var canAddToProject = function(rules) {
  return _.some(rules, function(rule) {
    return _.some(rule.resources, function(resource) {
      return checkResource(resource) && !_.isEmpty(_.intersection(rule.verbs ,(["*", "create", "update"])));
    }
  }
}
```